### DISCUSSION TITLE
How to test the instrumentation in an application using Golang test?

### DISCUSSION THREAD

**Discussion by ethernoy** [open-telemetry/opentelemetry-go#2891] [Q&A] [answered]
Given that we have a application that has three functions A, B and C, where A always calls B, B always calls C and C calls nothing else. In each of the function, a span is created using opentelemetry-go instrumentation. 

How can we correctly use go test to test that in one call of A, all three spans are created correctly? I noticed that we may use a otlp-collector to serve as a span collector, but I believe this is quite an overkill.

**Comment by dmathieu** [comment, marked as answer]
You can use the provider [oteltest](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace/tracetest) package, which includes an InMemory Span Exporter.
Configure your tracer provider with that exporter, run your code. And all spans will be available with the `GetSpans` method.

**Reply by ethernoy** [reply]
Thank you so much. One thing to notice though: better use `WithSyncer` instead of `WithBatcher` when setting the exporter in `TracerProvider`, otherwise the `GetSpan` call may have inconsistent response.

**Reply by dmathieu** [reply]
Obviously, yes.

**Reply by sdavispluto** [reply]
Why was oteltest removed? The deprecation message says to use tracetest, but that module doesn't include metrics. Is there another way to test metrics, now that oteltest is removed?

**Reply by dmathieu** [reply]
There is metricdatatest, which allows comparison of collected metrics: https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric/metricdata/metricdatatest
